networkit/cpp/matching/BMatching.cpp

```diff
@@ -0,0 +1,111 @@
+#include <networkit/matching/BMatching.hpp>
+
+namespace NetworKit {
+
+BMatching::BMatching(const std::vector<count> &b, count z)
+    : data(z, std::vector<node>(b.size(), none)), b(b) {}
+
+bool BMatching::isProper(const Graph &G) const {
+    // check if entries are symmetric
+    for (node v : G.nodeRange()) {
+        for (index i = 0; i < b.at(i); i++) {
+            node w = data[v][i];
+            if (w != none && std::find(data[w].begin(), data[w].end(), v) == data[w].end()) {
+                DEBUG("node ", v, " is not symmetrically matched");
+                return false;
+            }
+        }
+    }
+
+    // check if every pair exists as an edge
+    for (node v : G.nodeRange()) {
+        for (index i = 0; i < b.at(i); i++) {
+            node w = data[v][i];
+            if ((v != w) && (w != none) && !G.hasEdge(v, w)) {
+                DEBUG("matched pair (", v, ",", w, ") is not an edge");
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void BMatching::match(node u, node v) {
+    {
+        auto i = findFirstFreeIndex(u);
+        assert(i != none);
+        data[u][i] = v;
+    }
+    {
+        auto i = findFirstFreeIndex(v);
```

Comment:
There is a lot of `std::find` done in the code. Even though b is likely to be small, have you checked whether `std::set` (or a struct based on it) is a faster solution?

Reply:
Restructured BMatching in 14494fe204. Used `std::set` instead.